packages/react-components/react-toast/src/components/Toaster/useToaster.tsx

```diff
@@ -27,26 +27,38 @@ export const useToaster_unstable = (props: ToasterProps): ToasterState => {
 
   const rootProps = getNativeElementProps('div', rest);
 
-  const onFocusPositionSlot = useEventCallback((e: React.FocusEvent) => {
-    if (!e.currentTarget.contains(e.relatedTarget)) {
-      pauseAllToasts();
-    }
+  // Adds native HTML focusin/focusout listeners
+  // https://github.com/facebook/react/issues/25194
+  const focusListenerRef = React.useCallback(
+    (el: HTMLDivElement | null) => {
+      if (el) {
+        el.addEventListener('focusin', e => {
+          if (
+            isHTMLElement(e.currentTarget) &&
+            !e.currentTarget.contains(isHTMLElement(e.relatedTarget) ? e.relatedTarget : null)
+          ) {
+            pauseAllToasts();
+          }
+        });
 
-    rootProps.onFocus?.(e);
-  });
-
-  const onBlurPositionSlot = useEventCallback((e: React.FocusEvent) => {
-    if (!e.currentTarget.contains(e.relatedTarget)) {
-      playAllToasts();
-      tryRestoreFocus();
-    }
-
-    rootProps.onBlur?.(e);
-  });
+        el.addEventListener('focusout', e => {
```

Comment:
Is there no moment where you should remove this event listener?

Reply:
Once the HTMLElement is removed from DOM, the event listener is cleaned up with it. Generally it's  only the document level event listeners that are important to remove